Refactor request

weronika-kurczyna opened the pull request:
1. Made refactor in other endpoints (merge 2 functions in 1), reduce the code.
2. Deleted the alert message because it makes no sense. If the campaign doesn't exists, the script won't run and the error will occur.
3. Add clear information about missing campaign in main server.js, in console (error message) and the instruction in main readme and tiered-cart-promotion readme.

Commit:
refactor requests

weronika-kurczyna committed:
Change "attachEndpointsStackingPromotions" to "addStackingPromotionRoutes", remove "accessToStackingPromotionApp". Change names of "promotionStackableObj" and "defaultItems". Move "validateStackableParams" and "customer" objects to "addStackingPromotionRoutes" function.

weronika-kurczyna committed:
Add validateStackableParams and redeemStackableParams

Add validateStackableParams and redeemStackableParams instead of one global object

weronika-kurczyna committed:
Add check if campaign 'Reward Promotion' exists

weronika-kurczyna committed:
Add alert with information about lack of the campaign

weronika-kurczyna committed:
Reduction of the code

Assigning params directly in object.

weronika-kurczyna committed:
add cURL

weronika-kurczyna committed:
reformat cURL

weronika-kurczyna committed:
Update README.md

weronika-kurczyna committed:
Add information in server.js

weronika-kurczyna committed:
Delete the alert message

I deleted the alert because it started generating bugs when checking if the rewardPromotion exists.

weronika-kurczyna committed:
Refactor other endpoints

weronika-kurczyna committed:
Update README.md

marcin-slezak commented on README.md:
````diff
@@ -53,6 +53,227 @@ npm run start / npm run dev || yarn start / yarn run dev
+- Keep the returned validation_rule id after executing each request. It will be required for the main request.
+
+```
+curl --location --request POST 'https://api.voucherify.io/v1/validation-rules' \
````
As we discussed on Friday, a better option will be to write a small Nodejs CLI script, as then we can automate putting generated validation_rule id to campaign endpoint.

marcin-slezak replied:
here, you probably want to convince developer to use `missing-campaign.js` script instead of showing curls

weronika-kurczyna replied:
main README: no cURLs and
Note: The application should automatically add the missing campaign "Reward Promotion" by using script called "missing-campaign.js", so there is no need to add it manually.

README id tiered-cart-promotions:
Information previously added by Patryk + info about 'missing-campaign' + I left the cURL there "just in case", but I can also delete them from the second README.

marcin-slezak commented on server.js:
```diff
@@ -41,19 +41,31 @@ const checkCredentials = async () => {
     }
 };
+
+//Important!
```
It would be great to put a little attention to how we format code comments:

```ts
//Some comment

vs 

// Some comment with space before the first letter

//Multi line
//comment without a space at the beginning
// and starting with slashes 

vs

/*
  Multiline (block) comments with proper spacing
  make it easier to read.  
*/

```

weronika-kurczyna committed:
Add creating missing campaign JS file

weronika-kurczyna committed:
Change the structure of the functions

weronika-kurczyna committed:
Update server.js

weronika-kurczyna committed:
Update README.md

marcin-slezak commented on missing-campaign.js:
```diff
@@ -0,0 +1,205 @@
+    }
+
+    const createRewardPromotion = async () => {
+        await Promise.all([
```
You mixed two approaches:

chaining promises
```ts
const result1 = await somePromise1();
const result2 = await somePromise2();
const result3 = await somePromise3();
```

parallel promises
```ts

const results = await Promise.all([
  somePromise1(),
  somePromise2(),
  somePromise3()
])
```


You also mixed `async/await` syntax with `then` syntax.

And you miss promise handling for `client.campaigns.create`

At the end:


```ts

const validationRules = await Promise.all([
  client.validationRules.create(firstValidationRule),
  client.validationRules.create(secondValidationRule),
  client.validationRules.create(thirdValidationRule),
])

const rewardPromotion = createRewardPromotionObject(validationRules[0].id, validationRules[1].id, validationRules[2].id);

const response = await client.campaigns.create(rewardPromotion);

if (response.code !== 200) {
  return new Error;
}

```

weronika-kurczyna replied:
changed, but have to check one thing before pushing

marcin-slezak commented on tiered-cart-promotions/README.md:
````diff
@@ -24,6 +24,228 @@ The demo is running with a [Sandbox project](https://docs.voucherify.io/docs/tes
+- Keep the returned validation_rule id after executing each request. It will be required for the main request.
+
+```
+curl --location --request POST 'https://api.voucherify.io/v1/validation-rules' \
````
`missing-campaign.js` instead of curls?

marcin-slezak replied:
up

marcin-slezak replied:
We do not need to describe curl details if we have a JS script. Just leave a information that you need to run a script if you have missing campaigns

marcin-slezak commented on server.js:
```diff
@@ -24,36 +25,59 @@ app.use(bodyParser.urlencoded({ extended: true }));
+        console.log("The Reward Campaign was successfully created.");
+      }
+      catch (error) {
+        const msg = "The 'Reward Promotion' campaign not found. This campaign is required for 'tiered-cart-promotion' to work properly. \r\nPlease create a 'Reward Promotion' campaign first. \r\nYou can check the details by visiting: https://github.com/voucherifyio/voucherify-examples/tree/main/tiered-cart-promotions#creating-a-reward-promotion-campaign";
```
suggest `missing-campaign.js` ?

weronika-kurczyna committed:
Change readme and main server.js file

more info about 'missing-campaign' script

weronika-kurczyna committed:
Change the structure and naming a bit

marcin-slezak commented on README.md:
````diff
@@ -53,6 +53,8 @@ npm run start / npm run dev || yarn start / yarn run dev
 ```
 7. Go to [http://localhost:3000](http://localhost:3000/) in your browser.
 
+Note: The application should automatically add the missing campaign "Reward Promotion" by using script called "missing-campaign.js", so there is no need to add it manually.
````
`node missing-campaign`?  Also, it would be great to add a verb into the script name as we do some action, so maybe `node add-missing-campaign`?

marcin-slezak commented on missing-campaign.js:
```diff
@@ -0,0 +1,218 @@
+        "context_type": "campaign.promotion.discount.apply_to_order"
+    }
+
+    // const createRewardPromotion = async () => {
```
do not leave the code that is commented out

weronika-kurczyna committed:
update missing-campaign

weronika-kurczyna committed:
naming change

weronika-kurczyna committed:
Change the way errors are displayed

From now if an error from V% occurs, it will be logged to the console, but also the informative "msg" defined by me.

marcin-slezak commented on add-missing-campaign.js:
```diff
@@ -0,0 +1,207 @@
+export const addMissingCampaign = async (client) => {
+
+    const createRewardPromotionObject = (firstValidationRule, secondValidationRule, thirdValidationRule) => {
```
We do not have typescript support here, so even more important is to pay attention to naming. Please note that the `createRewardPromotionObject` function accepts as parameters id of three validation rules, not validation rule definitions from below. Therefore, the `firstValidationRule` name has two meanings in that file, once its rule id, and once its rule definition. I suggest changing a name from `firstValidationRule` to `firstValidationRuleId`

marcin-slezak commented on add-missing-campaign.js:
```diff
@@ -0,0 +1,207 @@
+export const addMissingCampaign = async (client) => {
```
I like the idea of using closure to provide V% client 👍 Please note that only the `addValidationRulesAndCampaign` function needs that context. The `createRewardPromotionObject`, `firstValidationRule`, `secondValidationRule` and `thirdValidationRule` functions do not need this context so they can be put outside of the outer function.

marcin-slezak commented on server.js:
```diff
@@ -24,36 +25,61 @@ app.use(bodyParser.urlencoded({ extended: true }));
 };
+
+/*
+Important!
```
This comment is outdated.

marcin-slezak commented on server.js:
```diff
@@ -24,36 +25,61 @@ app.use(bodyParser.urlencoded({ extended: true }));
+for getting instructions on how to create the missing campaign.
+*/
+
+const checkCampaign = async () => {
```
The `checkCampaign` function actually adds a campaign to the V% account if it does not exists, which may be a little dangerous if someone by accident connects the production account. I suggest does not allow to run the example (like for `checkCredentials`) and displaying the message that you have already in the catch block.

weronika-kurczyna committed:
Refactor - move functions, change naming and readme

- Changed the naming from "firstValidationRule" to "firstValidationRuleId" etc.
- Deleted unnecessary comments and text in Readme
- Formatted code with tslint formatter
- Moved "createRewardPromotionObject" and validation rules object outside the outer function

weronika-kurczyna committed:
Change the way of running "add-missing-campaign.js" script

weronika-kurczyna committed:
Little change in message

weronika-kurczyna committed:
Change format of the message in server.js (better readability)

marcin-slezak commented on .vscode/launch.json:
```diff
@@ -0,0 +1,17 @@
+{
```
This file should be excluded from the repo

marcin-slezak commented on server.js:
```diff
@@ -41,19 +41,27 @@ const checkCredentials = async () => {
+        await client.campaigns.get("Reward Promotion");
+    } catch (error) {
+        if (error.code === 404) {
+                const msg = "The 'Reward Promotion' campaign not found. \r\nPlease run the command 'node ./add-missing-campaign.js' in the terminal to create the missing campaign. \r\nWhen you receive a message that the campaign has been successfully created you can run the main script again. \r\nYou can get additional support here: https://github.com/voucherifyio/voucherify-examples/tree/main#ge ...
```
It's a little hard to read it. Also, \r\n is Windows-specific. Maybe:

```ts
const { EOL } = require('os');
 // ...
const msgs = [
  'The \'Reward Promotion\' campaign not found',
  'Please run the command \'node ./add-missing-campaign.js\' in the terminal to create the missing campaign.',
  'When you receive a message that the campaign has been successfully created you can run the main script again.',
  'You can get additional support here: https://github.com/voucherifyio/voucherify-examples/tree/main#get-support'
]

// try
console.log(msgs.join(EOL))
// or
console.log(...msgs)
```

weronika-kurczyna replied:
Thanks for the suggestion to use EOL! 🙂

weronika-kurczyna replied:
changes have been made

weronika-kurczyna committed:
Change the formatting of the console message using EOL

Deleted unnecessary file
